### PR TITLE
Switch over to host interface for any DNS error

### PR DESCRIPTION
**Pull request by dims**
(EDIT: incorporated https://github.com/kubernetes/kubernetes/pull/118927 into this one as it will merge first)

Not just then `IsNotFound`, but also for others like we see in the test suite

https://storage.googleapis.com/kubernetes-jenkins/logs/ci-cgroupv2-containerd-node-arm64-e2e-serial-ec2/1673979189503463424/build-log.txt
```
            Err: <*net.DNSError | 0x4000ba1e00>{
                Err: "server misbehaving",
                Name: "i-09c274ca42e0d4fbf",
                Server: "127.0.0.53:53",
                IsTimeout: false,
                IsTemporary: true,
                IsNotFound: false,
            }
```

#### What type of PR is this?
/kind cleanup

#### Which issue(s) this PR fixes:

Fixes #

#### Does this PR introduce a user-facing change?

```release-note
NONE
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs

```

### CODE REVIEW

**Comment by aojea** [comment]
some dns errors are fatal as "no try again, that record does not exist" and others should be retriable as "I'm busy now, please give me a sec and try again", assuming all of them are the former does not look right

**Comment by dims** [comment]
@aojea That was the problem i am trying to fix LOL

The test just fails when there is a DNS error (when making a http call), so i am adding code to do a DNS lookup before the http call so we can fallback to a known ip address when we know the DNS is not working temporarily

**Comment by dims** [comment]
@aojea see https://github.com/kubernetes/kubernetes/pull/118910

**Comment by aojea** [comment]
you can check if the DNS errors is temporary and retry https://pkg.go.dev/net#DNSError.Temporary

File: test/e2e_node/util.go [hunk 1]
```diff
@@ -305,7 +305,7 @@ func logKubeletLatencyMetrics(ctx context.Context, metricNames ...string) {
 	for _, key := range metricNames {
 		metricSet.Insert(kubeletmetrics.KubeletSubsystem + "_" + key)
 	}
-	metric, err := e2emetrics.GrabKubeletMetricsWithoutProxy(ctx, fmt.Sprintf("%s:%d", nodeNameOrIP(), ports.KubeletReadOnlyPort), "/metrics")
+	metric, err := e2emetrics.GrabKubeletMetricsWithoutProxy(ctx, fmt.Sprintf("%s:%d", nodeNameOrIP(ctx), ports.KubeletReadOnlyPort), "/metrics")
```
Comment (aojea) [inline test/e2e_node/util.go:308]:
```suggestion
        host := net.JoinHostPort(nodeNameOrIP(ctx), strconv.Itoa(ports.KubeletReadOnlyPort))
	metric, err := e2emetrics.GrabKubeletMetricsWithoutProxy(ctx, host, "/metrics")
```

Comment (SergeyKanzhelev) [reply test/e2e_node/util.go:308]: good idea to use `net.JoinHostPort` and `ports.KubeletReadOnlyPort` in tests above as well

File: test/e2e_node/util.go [hunk 2]
```diff
@@ -640,19 +640,26 @@ func WaitForPodInitContainerToFail(ctx context.Context, c clientset.Interface, n
...
-func nodeNameOrIP() string {
+func nodeNameOrIP(ctx context.Context) string {
 	// Check if the node name in test context can be resolved
-	if ips, err := net.LookupIP(framework.TestContext.NodeName); err != nil {
-		if dnsErr, ok := err.(*net.DNSError); ok && dnsErr.IsNotFound {
-			// if it can't be resolved, pick a host interface
-			if ip, err := utilnet.ChooseHostInterface(); err == nil {
-				return ip.String()
+	attempts := 10
+	for i := 0; i < attempts; i++ {
```
Comment (SergeyKanzhelev) [inline test/e2e_node/util.go:646]: should we sleep a bit between attempts?

File: test/e2e_node/util.go [hunk 3]
```diff
@@ -640,19 +640,26 @@ func WaitForPodInitContainerToFail(ctx context.Context, c clientset.Interface, n
 	})
 }
 
-func nodeNameOrIP() string {
+func nodeNameOrIP(ctx context.Context) string {
```
Comment (SergeyKanzhelev) [inline test/e2e_node/util.go:643]: same comment as https://github.com/kubernetes/kubernetes/pull/118927#discussion_r1245498641 - it started being overly complex comparing to just using the `localhost`

Comment (dims) [reply test/e2e_node/util.go:643]: Done! https://github.com/kubernetes/kubernetes/pull/118949

**Pull request closed without merging**